### PR TITLE
feat: implement cleanup extension point to the scheduling framework

### LINKED ISSUE #81438: Propose a cleanup extension point to the scheduling framework
/sig scheduling
/priority important-soon
/assign

**Why is this needed**:

The scheduling framework plugin could save pod related states in some extension points and used in others, ex:

1. [Save](https://github.com/kubernetes/kubernetes/pull/80898/files#diff-f58124514ebdc62aefca3f447c95bdadR59-R60) the pod to podAffinityPriorityMap in the Score extension point 
2. [Use](https://github.com/kubernetes/kubernetes/pull/80898/files#diff-f58124514ebdc62aefca3f447c95bdadR157-R160) and [delete](https://github.com/kubernetes/kubernetes/pull/80898/files#diff-f58124514ebdc62aefca3f447c95bdadR191) the map in the Normalize Score extension point.

If nothing goes wrong, the plugin would save, use and delete as usual. However, if the other score plugins return an `Error` status, the generic scheduler would return immediately without any callback to the plugin.

We could meet one of these problems:

1. If the pod was deleted and never scheduled again, the `podAffinityPriorityMap` will be saved in the plugin forever.
2. If the scheduler schedules the pod again, the plugin will use the data calculated in the last time. Though we could use a hack for this, ex: check whether the data was initialized before, if it was, delete the original data and initialize a new one, it does not sound like a good idea.

These problems happen because of:

1. There is no cleanup phase which notifies the plugin of the scheduling result (success or failure).
2. Scheduling the same pod could be a different scheduling process, the same extension point of the plugin could be called with multiple times with the same pod.

**What would you like to be added**:

We could solve these problems through the following approaches.

1. Introducing a new cleanup phase to the plugin which would be called once after a successful or failed scheduling.

```go
type CleanupPlugin interface {
	Plugin
	Cleanup(pc *PluginContext, p *v1.Pod)
}
```

2. Add a `cleanup` method to the `Plugin` interface

```go
type Plugin interface {
	Name() string
	Cleanup(pc *PluginContext, p *v1.Pod)
}
```

cc/ @bsalamat @Huang-Wei @ahg-g

### PR DESCRIPTION
**Pull request by draveness**
/kind feature
/sig scheduling
/priority important-soon
/hold
/assign @bsalamat @ahg-g 

**What this PR does / why we need it**:

Implement cleanup extension point to the scheduling framework, #81438 is created for the cleanup extension point discussion.

**Which issue(s) this PR fixes**:

Fixes #81438

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Implement "cleanup" extension point to the scheduling framework
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/apis/config/types.go [hunk 1]
```diff
@@ -164,6 +164,9 @@ type Plugins struct {
 	// NormalizeScore is a list of plugins that should be invoked after the scoring phase to normalize scores.
 	NormalizeScore *PluginSet
 
+	// Cleanup is a list of plugins that should be invoked after the scheduling framework is completed or aborted.
```
Comment (Zyqsempai) [inline pkg/scheduler/apis/config/types.go]: We have a documentation convention, so it should be:
```suggestion
	// `cleanup` is a list of plugins that should be invoked after the scheduling framework is completed or aborted.
```
Please. also fix for other type properties.

Comment (draveness) [reply pkg/scheduler/apis/config/types.go]: I don't think this is valid, please take a look at the other comments in this file.

Comment (Zyqsempai) [reply pkg/scheduler/apis/config/types.go]: Yep, sorry, my bad, it's valid only for json/grpc tagged types, sorry.

**Pull request closed without merging**